frontend/src/metabase/admin/datamodel/containers/MetricApp.jsx

```diff
@@ -6,73 +6,86 @@ import MetabaseAnalytics from "metabase/lib/analytics";
 
 import MetricForm from "./MetricForm.jsx";
 
-import { metricEditSelectors } from "../selectors";
 import * as actions from "../datamodel";
-import { clearRequestState } from "metabase/redux/requests";
-import { fetchTableMetadata } from "metabase/redux/metadata";
 import { getMetadata } from "metabase/selectors/metadata";
+import Metrics from "metabase/entities/metrics";
+import Tables from "metabase/entities/tables";
 
 const mapDispatchToProps = {
   ...actions,
-  fetchTableMetadata,
-  clearRequestState,
+  createMetric: Metrics.actions.create,
   onChangeLocation: push,
 };
 
 const mapStateToProps = (state, props) => ({
-  ...metricEditSelectors(state, props),
   metadata: getMetadata(state, props),
 });
 
-@connect(
-  mapStateToProps,
-  mapDispatchToProps,
-)
-export default class MetricApp extends Component {
-  async componentWillMount() {
-    const { params, location } = this.props;
+@Metrics.load({
+  id: (state, props) => parseInt(props.params.id),
+  wrapped: true,
+})
+@Tables.load({ id: (state, props) => props.metric.table_id, wrapped: true })
+class UpdateMetricForm extends Component {
+  constructor() {
+    super();
+    this.onSubmit = this.onSubmit.bind(this);
+  }
 
-    let tableId;
-    if (params.id) {
-      const metricId = parseInt(params.id);
-      const { payload: metric } = await this.props.getMetric({ metricId });
-      tableId = metric.table_id;
-    } else if (location.query.table) {
-      tableId = parseInt(location.query.table);
-    }
+  async onSubmit(metric) {
```

Comment:
We've started using the class properties syntax for binding, e.x.

```
onSubmit = async (metric) => {
```

Then you can get rid of the constructor.

Reply:
BTW, only do that for methods you're passing as props or otherwise need it to be bound since it's a bit slower.